docs/use_cases/real_time_apps/part2.md

```diff
@@ -0,0 +1,366 @@
+- **Start a web server** with three API endpoints:
+    * GET /api/stats
+    * GET /api/spreadsheets
+    * POST /api/spreadsheets
+
+We’ll focus on the spreadsheet endpoints and skip stats because it’s similar to fetching cell data (except it sends spreadsheet statistics).
+
+### Sending Cell Data: `GET /api/spreadsheets`
+
+This endpoint provides an initial snapshot of the requested Feldera spreadsheet cells, then streams any subsequent changes within that range to the client.
+
+We use a WebSocket endpoint for a bidirectional connection, following a simple protocol:
+
+1. **Client request** – The client requests a range of cells, sending for example: e.g., `{ from: 0, to: 4 }`
+
+2. **Initial response** – The server returns the cell contents in ndjson format:
+
+    ```json
+    { id: 0, raw_value: 41, computed_value: 41 }
+    { id: 1, raw_value: 1, computed_value: 1 }
+    { id: 2, raw_value: '=B0', computed_value: 1 }
+    { id: 3, raw_value: '=A0+C0', computed_value: 41 }
+    ```
+
+3. **Updates** – If cells within that range change, the server sends incremental updates:
+
+    ```json
+    { id: 0, raw_value: 0, computed_value: 0 }
+    { id: 3, raw_value: 1, computed_value: 1 }
+    ```
+
+4. **Range changes** – The client can request a different range by sending a new JSON line with updated `from` and `to` values, receiving a fresh snapshot and corresponding updates.
+
+The implementation resides in [spreadsheet.rs](https://github.com/feldera/techdemo-spreadsheet/blob/5abfb3aedc1ffa38b23341df2ed951726aef86f0/server/src/spreadsheet.rs#L46). It includes exhaustive error handling, so we’ll only show key snippets here.
+
+The HTTP handler is brief: it upgrades the connection to WebSocket and calls `.subscribe()` on `xls_subscription` which listens to changes from Feldera’s spreadsheet_view.
+
+```rust
+pub(crate) async fn ws_handler(
+    ws: WebSocketUpgrade,
+    ConnectInfo(addr): ConnectInfo<SocketAddr>,
+    State(state): State<AppState>,
+) -> impl IntoResponse {
+    debug!("{addr} connected.");
+    ws.on_upgrade(move |socket| handle_socket(state.xls_subscription.subscribe(), socket, addr))
+}
+```
+
+Let's look at the main logic which is in `handle_socket`. This function internally spawns **two asynchronous tasks**:
+
+- **Client messages** – The first task listens for messages from the client in the form `{ from: x, to: y }`, indicating which cells the client wants. When such a request arrives, we send an [ad-hoc query](https://docs.feldera.com/api/execute-an-ad-hoc-query-in-a-running-or-paused-pipeline) to Feldera to fetch that range and forward the result to the client:
+
+```rust
+match process_message(msg, who) {
+ControlFlow::Continue(Some(region)) => {
+    match adhoc_query(
+        format!(
+            "SELECT * FROM spreadsheet_view WHERE id >= {} and id < {}",
```

Comment:
We don't use indices for these queries right?

Reply:
No